### PR TITLE
Improve link handling in markdown links

### PR DESCRIPTION
**Pull request by martin-fleck-at**
#### What it does

- Extend EditorManager to have more flexible selection resolution
- Allow adopters to register selection resolvers
-- Provide selection resolver for GitHub location syntax
-- Provide selection resolver for TypeDoc symbol navigation
-- Provide selection provider for fragment textual search
- Fix minor issue in editor-manager where URI is not passed properly

Fixes https://github.com/eclipsesource/osweek-2024/issues/18

#### How to test

- GitHub Location match: Ensure that you generate a link with `file://<path>#<location>` where location is `L<start-line>C<start-column>-L<end-line>C<end-column>` (end position is optional and all columns are optional as well)
- Symbol match: Ensure that you generate a link with `file://<path>#<symbol-path>` where symbol path can be just the symbol name (e.g., a function name in TypeScript) or a nested path in the form of `parent.child` (e.g., a function nested in a namespace). Ensure symbols are actually provided for your document by checking the Outline or triggering the command 'Go to Symbol...` (not Go to Symbol in Workspace...).
- Any text match: Ensure that you generate a link with `file://<path>#<text>`

link-support

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Review by lucas-koehler — COMMENTED** [review]
Code looks pretty good to me and works well to me. I have only two tiny remarks/questions inline :)

File: packages/ai-chat-ui/src/browser/chat-response-renderer/ai-editor-manager.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                    character: matches[0].range.end.character - 1
+                }
+            };
+        }
+    }
+
+    protected findFragment(uri: URI): string | undefined {
+        const input = uri.toString();
+        const hashIndex = input.indexOf('#');
+        return hashIndex === -1 ? undefined : input.substring(hashIndex + 1);
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/ai-chat-ui/src/browser/chat-response-renderer/ai-editor-manager.ts:155]: Good catch! I was starting with string manipulation and only later moved it into the editor manager where we already have the URI, thank you!

File: packages/ai-chat-ui/src/browser/chat-response-renderer/ai-editor-manager.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            if (match) {
+                return this.m2p.asRange(match.selectionRange);
+            }
+        }
+    }
+
+    protected findSymbolPath(uri: URI): string[] | undefined {
+        const input = uri.toString();
+        const hashIndex = input.indexOf('#');
+        return hashIndex === -1 ? undefined : input.substring(hashIndex + 1).split('.');
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/ai-chat-ui/src/browser/chat-response-renderer/ai-editor-manager.ts:107]: Same as above, good catch, thank you!

**Review by lucas-koehler — APPROVED** [review]
LGTM thanks :sparkles:

**Pull request merged**